d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/RelativeStateUpdater.java

```diff
@@ -0,0 +1,427 @@
+  }
+
+  RelativeStateUpdater(D2RelativeStrategyProperties relativeStrategyProperties,
+      QuarantineManager quarantineManager,
+      ScheduledExecutorService executorService,
+      ConcurrentMap<Integer, PartitionState> partitionLoadBalancerStateMap,
+      List<PartitionStateUpdateListener.Factory<PartitionState>> listenerFactories)
+  {
+    _relativeStrategyProperties = relativeStrategyProperties;
+    _quarantineManager = quarantineManager;
+    _executorService = executorService;
+    _listenerFactories = listenerFactories;
+    _partitionLoadBalancerStateMap = partitionLoadBalancerStateMap;
+    _lock = new ReentrantLock();
+
+    _executorService.scheduleWithFixedDelay((Runnable) this::updateState, EXECUTOR_INITIAL_DELAY,
+        _relativeStrategyProperties.getUpdateIntervalMs(), TimeUnit.MILLISECONDS);
+  }
+
+  @Override
+  public void updateState(Set<TrackerClient> trackerClients, int partitionId, long clusterGenerationId)
+  {
+    if (!_partitionLoadBalancerStateMap.containsKey(partitionId))
+    {
+      // If the partition is not initialized, initialize the state synchronously
+      initializePartition(trackerClients, partitionId, clusterGenerationId);
+    }
+    else if (clusterGenerationId != _partitionLoadBalancerStateMap.get(partitionId).getClusterGenerationId())
+    {
+      // Asynchronously update the state if it is from uri properties change
+      PartitionState oldPartitionState = _partitionLoadBalancerStateMap.get(partitionId);
+      _executorService.execute(() -> updateStateForPartition(trackerClients, partitionId, oldPartitionState, clusterGenerationId));
+    }
+  }
+
+  @Override
+  public Ring<URI> getRing(int partitionId)
+  {
+    return _partitionLoadBalancerStateMap.get(partitionId) == null
+        ? null
+        : _partitionLoadBalancerStateMap.get(partitionId).getRing();
+  }
+
+  Map<URI, Integer> getPointsMap(int partitionId)
+  {
+    return _partitionLoadBalancerStateMap.get(partitionId) == null
+        ? new HashMap<>()
+        : _partitionLoadBalancerStateMap.get(partitionId).getPointsMap();
+  }
+
+  /**
+   * Update the partition state.
+   * This is scheduled by executor, and the update should not involve any cluster change
+   */
+  void updateState()
+  {
+    // Update state for each partition
+    for (Integer partitionId : _partitionLoadBalancerStateMap.keySet())
+    {
+      PartitionState partitionState = _partitionLoadBalancerStateMap.get(partitionId);
+      if (partitionState != null)
+      {
+        updateStateForPartition(partitionState.getTrackerClients(), partitionId, partitionState, null);
+      }
+    }
+  }
+
+  /**
+   * Update the partition state when there is a cluster uris change.
+   */
+  void updateStateForPartition(Set<TrackerClient> trackerClients, int partitionId,
+      PartitionState oldPartitionState, Long clusterGenerationId)
+  {
+    LOG.debug("Updating for partition: " + partitionId + ", state: " + oldPartitionState);
```

Comment:
I took a look into the implementation of LogUtil.debug, found it does nothing but just a check of log.isDebugEnabled(). Then I explicitly checked log4j implementation, it definitely takes care of the log level config too. So I just changed to use LOG directly since it's straightforward.

Reply:
Double checked on log4j implementation, it has exact same implementation of log level check and string array arguments. Will keep using log4j implementation directly.